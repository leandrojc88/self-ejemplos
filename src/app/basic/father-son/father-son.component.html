<h5>Hijo</h5>

<code>{{ client | json }}</code>

<br /><br />

<button
  *ngIf="client"
  class="btn btn-primary"
  (click)="onUpdate(5)"
  data-test="btn-update"
>
  Cambiar ID
</button>

<button
  *ngIf="client"
  class="btn btn-outline-danger"
  (click)="onDelete()"
  data-test="btn-delete"
>
  Borrar Cliente
</button>
