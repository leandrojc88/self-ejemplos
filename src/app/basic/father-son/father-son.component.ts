import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Client } from '../model';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss'],
})
export class FatherSonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onUpdatedClient = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onUpdate(id: number) {
    if (!this.client) return;

    this.client = { ...this.client, id };
    this.onUpdatedClient.emit(this.client);
  }
}
