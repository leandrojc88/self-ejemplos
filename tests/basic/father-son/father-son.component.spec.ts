import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [FatherSonComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('shold by match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('shold not view buttons if not a plient', () => {
    const button = compiled.querySelectorAll('button');
    expect(button.length).toBe(0);
  });

  it('shold view buttons if plient exist', () => {
    component.client = { id: 1, name: 'LEO' };
    fixture.detectChanges();

    const button = compiled.querySelectorAll('button');
    expect(button.length).toBe(2);
  });

  it('shold by match snapshot if client exist', () => {
    component.client = { id: 1, name: 'LEO' };
    fixture.detectChanges();
    expect(compiled).toMatchSnapshot();
  });

  it('should emit onDeleteClient when delete button is clicked', () => {
    component.client = { id: 1, name: 'LEO' };
    fixture.detectChanges();

    // esto es como un mock de una funcion
    jest.spyOn(component.onDeleteClient, 'emit');
    const buttonDelete = compiled.querySelector('[data-test=btn-delete]');

    buttonDelete?.dispatchEvent(new Event('click'));
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });

  it('should emit onUpdatedClient when "Cambiar ID" button is clicked', () => {
    component.client = { id: 1, name: 'LEO' };
    fixture.detectChanges();

    // esto es como un mock de una funcion
    jest.spyOn(component.onUpdatedClient, 'emit');
    const buttonUpdate = compiled.querySelector('[data-test=btn-update]');

    buttonUpdate?.dispatchEvent(new Event('click'));
    expect(component.onUpdatedClient.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'LEO',
    });
  });

  it('should emit onUpdatedClient if "client" is exist ', () => {
    // esto es como un mock de una funcion
    jest.spyOn(component.onUpdatedClient, 'emit');
    component.onUpdate(10);
    expect(component.onUpdatedClient.emit).not.toHaveBeenCalled();

    component.client = { id: 1, name: 'LEO' };
    fixture.detectChanges();
    component.onUpdate(10);
    expect(component.onUpdatedClient.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'LEO',
    });
  });
});
