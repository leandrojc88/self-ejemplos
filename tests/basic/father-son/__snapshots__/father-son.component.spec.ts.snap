// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent shold by match snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;

exports[`FatherSonComponent shold by match snapshot if client exist 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "LEO"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-primary"
    data-test="btn-update"
  >
     Cambiar ID

  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
     Borrar Cliente

  </button>
  
</div>
`;
